### PR TITLE
fix(augmentation): draw each CutMix cut's placement per row and mix

### LINKED ISSUE #4712: RandomCutMixV2: cut placement is one uniform draw shared by every row and mix, even with same_on_batch=False (regression from #1459)
## Describe the bug

`CutmixGenerator.forward` draws **one** uniform number per axis for the cut start position and shares it across every row and every mix of the batch, even with `same_on_batch=False`. Only the cut *sizes* are drawn per row. The placement of all `B * num_mix` cuts is therefore perfectly correlated: every cut sits at the same relative position `u * (side - cut_side - 1)`.

```python
_gen_shape = (1,)
if same_on_batch:
    _gen_shape = (cut_height.size(0),)
    ...
x_start = _adapted_rsampling(_gen_shape, self.rand_sampler, same_on_batch) * (width - cut_width - 1)
```

`rand_sampler` is a scalar `Uniform(0, 1)`, so `_gen_shape = (1,)` yields a single draw that is broadcast against the per-row `cut_width`. With `same_on_batch=True`, `_adapted_rsampling` also repeats one draw, so both modes use one placement draw; they differ only in the sizes.

## Reproduction

`main` @ `c90c05f66`, torch 2.14.0, CPU, macOS arm64:

```python
import torch, kornia.augmentation as K
from kornia.geometry.bbox import infer_bbox_shape

H = W = 100000  # large, so floor() is negligible
torch.manual_seed(0)
aug = K.RandomCutMixV2(p=1.0, num_mix=2, same_on_batch=False, use_correct_lambda=True)
c = aug.forward_parameters(torch.Size([4, 1, H, W]))["crop_src"].double().view(-1, 4, 2)
h, w = infer_bbox_shape(c)
print(w.tolist())
print((c[:, 0, 0] / (W - w - 1)).tolist())
print((c[:, 0, 1] / (H - h - 1)).tolist())
```

```
widths          [78942, 95768, 87808, 85240, 85751, 51559, 76970, 84918]      # per-row, as expected
x0 / (W-w-1)    [0.03400, 0.03380, 0.03396, 0.03401, 0.03397, 0.03402, 0.03400, 0.03402]   # one draw
y0 / (H-h-1)    [0.94420, 0.94422, 0.94422, 0.94424, 0.94420, 0.94424, 0.94424, 0.94423]   # one draw
```

With a fixed cut size (`cut_size=(0.5, 0.5)`) the consequence is that `crop_src` is identical across all 16 cuts of a `B=8, num_mix=2` batch in 200 of 200 seeds with `same_on_batch=False`, i.e. `same_on_batch` has no observable effect on the geometry there.

## Expected behavior

With `same_on_batch=False`, each row and mix draws its own start position, as it already draws its own size.

## Provenance: a regression, not the original design

The original implementation (#609, `fe6c54c95`) called `_adapted_uniform((1,), zeros_like(cut_width), width - cut_width, same_on_batch)` with **batch-shaped** `low`/`high`, so a sample shape of `(1,)` produced `(1, B * num_mix)` independent draws. The random-generator refactor (#1459, `fb7565a2f`) replaced that with a scalar `Uniform(0, 1)` sampler times a scale and kept `_gen_shape = (1,)`, which collapsed the placement to one draw.

## Suggested fix

Sample `(batch_size * num_mix,)` start positions when `same_on_batch=False`. This changes the RNG stream of `RandomCutMixV2` for a fixed seed, so it wants a `changelog.d` `fixed` fragment, and the doctest on `RandomCutMixV2` plus any seeded expected values need regenerating. A pin should use an input where sizes are fixed (`cut_size=(0.5, 0.5)`) so that only the placement can differ, and assert that not all `crop_src` rows are equal; the `same_on_batch=True` pin should use the default `cut_size` so that sizes are the discriminating part.

Found while reviewing #4711.

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).

### PR DESCRIPTION
**Pull request by thegoodengineer**
## What does this pull request do?

`RandomCutMixV2` draws the cut *size* once per row and mix, but the cut *placement* came from one uniform draw per axis that every cut in the batch shared, even with `same_on_batch=False`. All `B * num_mix` cuts therefore sat at the same relative position `u * (side - cut_side - 1)`. With a fixed `cut_size` the sixteen boxes of a `B=8, num_mix=2` batch were identical in 200 of 200 seeds, so `same_on_batch` had no visible effect on the geometry there.

The issue's reproduction, `B=4, num_mix=2, H=W=100000`, seed 0, before and after:

```text
                before (main)                                              after (this branch)
widths        [78942, 95768, 87808, 85240, 85751, 51559, 76970, 84918]   unchanged (sizes were already per cut)
x0 / (W-w-1)  [0.03400, 0.03380, 0.03396, 0.03401, 0.03397, 0.03402, ...]  [0.03400, 0.94422, 0.88016, 0.00122, 0.59356, 0.41575, 0.41769, 0.27107]
y0 / (H-h-1)  [0.94420, 0.94422, 0.94422, 0.94424, 0.94420, 0.94424, ...]  [0.69226, 0.20373, 0.68329, 0.75283, 0.85787, 0.68695, 0.00512, 0.17565]
```

`CutmixGenerator.forward` now samples `(batch_size * num_mix,)` start positions per axis, the shape the sizes already use. `_adapted_rsampling` repeats a single draw across that shape when `same_on_batch=True`, which is exactly what the old code requested for that case, so the shared path is untouched. The `_gen_shape = (1,)` came from the random-generator refactor in #1459, which replaced the batch-shaped `low`/`high` draw of the original implementation (#609) with a scalar `Uniform(0, 1)` sampler and kept the `(1,)` sample shape, collapsing the placement to one draw.

This changes the random stream of `RandomCutMixV2` for a fixed seed whenever `same_on_batch=False` and `B * num_mix > 1`, so it carries a `fixed` changelog fragment.

## Related issue or discussion

Closes #4712.

## Pins moved with the fix

- `test_wart_cutmix_placement_is_one_shared_draw_4712` becomes `test_convention_cutmix_placement_is_drawn_per_cut_4712`. It keeps the issue's suggested setup, `cut_size=(0.5, 0.5)` so that only the placement can differ, checks that the sizes really are pinned, then asserts that the rows of each mix land in different places and that the two mixes of a row differ. It also checks every box stays inside the image with the reserved pixel, so the `- 1` arithmetic is covered under the independent draws.
- `test_convention_cutmix_same_on_batch_shares_cut_geometry` is untouched; it already uses the default `cut_size` so that sizes are the discriminating part, as the issue asks.
- `TestRandomCutMixGen::test_random_gen` (seed 42, `B=8`, `p=0.5`, `same_on_batch=False`) pins a seeded literal. Its `crop_src` moved and is regenerated; `mix_pairs` is drawn before the starts and is unchanged. The old literal showed the defect directly: the five `p`-gated zero-size rows all sat at `[48, 31]`. The regenerated float32 and float64 outputs are identical. The bfloat16 cell of that test passed on `main` only because the shared draw happened to produce bfloat16-exact numbers; with independent draws, `floor(u * span)` under an 8-bit significand lands one pixel off the float32 literal in 3 of 8 rows, so that cell compares with `atol=1` and a comment says why. It is not a manifest entry.
- `TestRandomCutMixGen::test_same_on_batch` and every other seeded CutMix test still pass unchanged, including the docstring example, because they use `same_on_batch=True` or images small enough that `floor()` collapses the start to the same pixel.
- The Convention bullet on `RandomCutMixV2` no longer describes the shared placement and no longer cites #4712. `grep -rnE "#4712|issues/4712" kornia/` is empty; the only hit in `tests/` is the flipped pin's comment.

## What did you check?

**Nothing that shared a draw before changes.** I captured 540 `CutmixGenerator` outputs on `main` and on this branch: batch shapes `(0, 8, 8)`, `(1, 8, 8)`, `(2, 3, 4)`, `(8, 64, 48)` and `(5, 17, 33)`; `num_mix` 1, 2 and 3; `cut_size` `None`, `(0.5, 0.5)` and `(0.2, 0.8)`; both `same_on_batch` values; RNG dtypes float32 and float64; three seeds. Each probe ran from its own tree's root with `kornia.__file__` printed to confirm which tree it imported.

```text
same_on_batch=True                          270 / 270 bitwise identical
same_on_batch=False, B * num_mix <= 1        72 /  72 bitwise identical (one cut is one draw either way)
same_on_batch=False, B * num_mix >  1        61 / 198 identical; all 61 are the (2, 3, 4) and (1, 8, 8) shapes,
                                                       where floor() collapses the start to the same pixel
mix_pairs, output keys, shapes and dtypes   identical in all 540
```

**The moved pins fail on `main`,** run from a worktree of `main` with the updated test files:

```text
3 failed, 686 passed
FAILED test_convention_cutmix_placement_is_drawn_per_cut_4712   - assert 1 > 1  (rows of one mix all in one place)
FAILED TestRandomCutMixGen::test_random_gen[cpu-float32]        - Mismatched elements: 60 / 64
FAILED TestRandomCutMixGen::test_random_gen[cpu-float64]        - Mismatched elements: 60 / 64
```

**Compile.** `CutmixGenerator.forward` traces with `fullgraph=True` on the eager backend for both `same_on_batch` values, and the compiled parameters are `torch.equal` to eager. This machine has no MSVC, so `inductor` cannot build here; CI's dynamo leg is the one to confirm it.

```text
$ pytest tests/augmentation/test_conventions_mix.py tests/augmentation/test_random_generator.py \
      tests/augmentation/test_augmentation_mix.py -k "CutMix or Cutmix or cutmix or Mix" --dtype=float32,float64
1180 passed, 917 deselected

$ ... --dtype=float16  --xfail-known-failures --known-failure-profile=cpu-float16
505 passed, 108 xfailed          (the Beta sampler's float16 limits, all already in the manifest)

$ ... --dtype=bfloat16 --xfail-known-failures --known-failure-profile=cpu-bfloat16
611 passed, 1 xfailed

$ pytest tests/augmentation/test_augmentation_mix.py tests/augmentation/container tests/augmentation/test_base.py \
      tests/augmentation/test_conventions_mix.py tests/augmentation/test_random_generator.py tests/geometry/test_bbox.py --dtype=float32
2043 passed, 233 skipped, 7 xfailed

$ pytest --doctest-modules kornia/augmentation/_2d/mix/cutmix.py kornia/augmentation/random_generator/_2d/cutmix.py
1 passed

$ pre-commit run --files <the four changed files>
all hooks passed
```

I did not run CUDA or MPS locally.

### CODE REVIEW

**Review by ducha-aiki — APPROVED** [review]
Review of `8e45549cc`, plus the maintainer fix `6515d5b71`

This PR draws the CutMix cut placement once per row and mix (`_gen_shape = (batch_size * num_mix,)`) and does what #4712 asked: per-cut starts when `same_on_batch=False`, the wart pin flipped with `cut_size=(0.5, 0.5)` so only the placement can differ, the `same_on_batch=True` pin left on the default `cut_size`, the seeded literal regenerated, the Convention bullet rewritten, and a `fixed` fragment added. The library change is right. The one gap was in the flipped pin, and I fixed it myself (below).

**What I verified**
- `same_on_batch=True`: the old code already asked for `(cut_height.size(0),) == (B * num_mix,)` samples in that mode, and `_adapted_rsampling` repeats one draw, so that stream is unchanged. `test_same_on_batch` and `test_convention_cutmix_same_on_batch_shares_cut_geometry` pass without edits. `B * num_mix == 1` is `(1,)` both before and after.
- Run on `origin/main` library code, the new tests fail: `test_convention_cutmix_placement_is_drawn_per_cut_4712` (`assert 1 > 1`), plus `test_random_gen` in float32, float64 and bfloat16.
- Mutations of `random_generator/_2d/cutmix.py`, each killed by at least one test: `_gen_shape = (1,)` (254 red); `same_on_batch` → `True` in the x call alone or the y call alone (killed by `test_random_gen` f16/bf16/f32/f64); `same_on_batch` → `False` in both calls (killed by both `same_on_batch` pins); dropping the `- 1` on either axis (killed by `test_random_gen` f32/f64/f16).
- Image content: for `B=6, cut_size=(0.5, 0.5)`, each output row equals its input with exactly its own `crop_src` box pasted from its `mix_pairs` source. There were 5 distinct x0 and 6 distinct y0 values.
- dtype and precision: the draws keep `.to(device=_device, dtype=_dtype)`. The CutMix tests pass in float32/float64 (744), in float16 under `cpu-float16` (268 passed, 108 xfailed, all already in the manifest), and in bfloat16 under `cpu-bfloat16` (375 passed, 1 xfailed). The bfloat16 xfail is `test_same_on_batch[cpu-bfloat16]`, which is already in the manifest and fails the same way on main. The `atol=1` bfloat16 cell is justified, and float16 still matches the literal at 1e-4. The doctests pass.
- Compile: `torch._dynamo.explain(CutmixGenerator.forward)` gives 1 graph and 0 breaks for both `same_on_batch` values. The CutMix tests pass under `KORNIA_TEST_OPTIMIZER=inductor` (368).
- `grep -rnE "#4712|issues/4712" kornia/ tests/ docs/` finds only the flipped pin's comment. `conventions.rst` has no placement prose. `test_conventions_*.py` gives 694 passed on the head. On the merge with current `origin/main`, the conventions, random-generator and mix files give 3127 passed. The merge also passes with #4726 (`7e4f406`) on top: `test_conventions_mix.py` auto-merges and 412 tests pass.
- Changelog: a `fixed` fragment fits. This follows the #4610 precedent: a distribution defect is repaired, and no API contract changes.

**Flipped pin: fixed in `6515d5b71` (pushed by the maintainer)**

The pin's comment claimed "one pixel is reserved on each axis", but its bound `< shape[-1] - 1` held with or without the `- 1`. Its `(x, y)`-pair uniqueness check also passed if only one axis's draw were shared. Both of those mutants were caught only by the seeded literal in `test_random_gen`.

The reservation is not a contract worth pinning. The `- 1` plus the `floor` mean the last **two** legal starts per axis are never drawn, and `cut_size=(0.0, 0.0)` starts a full cut at `-1`. I filed that as #4730 instead of locking it in. The commit changes only this test:
- it adds a per-axis `unique().numel() > 1` check; a mutant sharing only the x draw, or only the y draw, now fails it;
- the bound is now "every corner lies inside the image" (`0 <= corner <= size - 1`), with the reservation claim dropped, so fixing #4730 will not need to touch this pin.

The CutMix tests in `test_conventions_mix.py` and `test_random_generator.py` give 352 passed, and ruff is clean. The PR body's sentence that the pin covers the `- 1` arithmetic is superseded by this.

**Verdict: approve** at `6515d5b71`. Nothing is left open; CI is the remaining gate.

Posted on behalf of @ducha-aiki by Claude (Opus 5).
